pkg/apis/kf/v1alpha1/app_types.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the License);
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an AS IS BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+import duckv1beta1 "github.com/knative/pkg/apis/duck/v1beta1"
+import "k8s.io/kubernetes/pkg/apis/core"
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// App is a 12-factor application deployed to Knative. It encompasses source
+// code, configuration, and the current state of the application.
+type App struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec AppSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status AppStatus `json:"status,omitempty"`
+}
+
+// AppSpec is the desired configuration for an App.
+type AppSpec struct {
+
+	// Source contains the source configuration of the App.
+	// +optional
+	Source AppSpecSource `json:"source,omitempty"`
+
+	// Template defines the App's desired runtime configuration.
+	// +optional
+	Template AppSpecTemplate `json:"template,omitempty"`
+
+	// Routes defines network routes for the App's ingress.
+	// +optional
+	Routes AppSpecRoutes `json:"routes,omitempty"`
+
+	// Services defines what services the App requires.
+	// +optional
+	Services AppSpecServices `json:"services,omitempty"`
+
+	// Serve determines if the App should be running.
+	Serve *bool `json:"serve"`
+}
+
+// AppSpecSource defines the source code for an App.
+// The fields ContainerImage and BuildpackBuild are mutually exclusive.
+type AppSpecSource struct {
+
+	// UId is a unique identifier for an AppSpecSource.
+	// Updating sub-values will trigger a new value.
+	// +optional
+	UId string `json:"uid,omitempty"`
+
+	// ContainerImage defines the container image for source.
+	// +optional
+	ContainerImage AppSpecSourceContainerImage `json:"containerimage,omitempty"`
```

Comment:
WDYT about using a struct here like what Knative Build has? https://github.com/knative/build/blob/master//pkg/apis/build/v1alpha1/build_types.go#L160:6

Alternatively, Tekton has the following: https://github.com/tektoncd/pipeline/blob/master/pkg/apis/pipeline/v1alpha1/resource_types.go#L70 

which is a kind of extensible way to do it because they support GCS, Git, Images, and some K8s types:

```
type PipelineResourceSpec struct {
	Type   PipelineResourceType `json:"type"`
	Params []Param              `json:"params"`
	// Secrets to fetch to populate some of resource fields
	// +optional
	SecretParams []SecretParam `json:"secrets,omitempty"`
}
```

If you merged the two maybe something like the following:

```
source:
  git:
    url: git@...
    revision: master 
  custom:
    # *corev1.Container
```

I think we'd only need the image for now, but it could be a quick win to get allow git -> executable.

Reply:
@josephlewis42 and I talked about this. We're going to keep things how they are now and re-evaluate when creating the reconciler